File: worthly-web/src/lib/merchant-rule.ts
export type MerchantRuleHint = {
  matchValue: string;
  operator: "CONTAINS" | "EQUALS";
  display: string;
  field: "MERCHANT" | "DESCRIPTION";
};

export type MerchantRuleLike = {
  field: string;
  operator: string;
  matchValue: string;
  enabled?: boolean;
};

const PREFIXES = [
  /^compra estrang\b/,
  /^compra mbw\b/,
  /^compra\b/,
  /^debito direto-?/,
  /^debito direto\b/,
  /^levantamento de numerario-?/,
  /^levantamento de numerario\b/,
  /^top-?up by\b/,
  /^top up by\b/,
];

const DROP = new Set([
  "lda",
  "sa",
  "unipessoal",
  "comercial",
  "servicos",
  "serviços",
  "portugal",
  "lisboa",
  "porto",
  "braga",
  "coimbra",
  "faro",
  "setubal",
  "aveiro",
  "pt",
  "the",
  "de",
  "do",
  "da",
  "dos",
  "das",
  "e",
  "com",
  "www",
  "bill",
  "inst",
  "mb",
  "atm",
  "sucursal",
  "loja",
  "compra",
  "estrang",
  "mbw",
]);

const GENERIC = new Set([
  "paypal",
  "revolut",
  "klarna",
  "eupago",
  "mbway",
  "mb way",
  "trading 212",
  "trading212",
  "apple pay",
  "google pay",
  "compra",
]);

export function stemMerchant(raw: string | null | undefined): MerchantRuleHint | null {
  if (!raw) {
    return null;
  }
  let text = normalizeMerchant(raw);
  if (!text) {
    return null;
  }
  let changed = true;
  while (changed) {
    changed = false;
    for (const prefix of PREFIXES) {
      const next = text.replace(prefix, "").trim();
      if (next !== text) {
        text = next;
        changed = true;
      }
    }
  }
  const tokens = text
    .split(" ")
    .map((token) => token.replace(/^\.+|\.+$/g, ""))
    .filter((token) => token.length >= 2 && !/^\d+$/.test(token) && !DROP.has(token));
  if (tokens.length === 0) {
    return null;
  }
  const matchValue =
    tokens.length >= 2 && tokens[0].length <= 6 && !tokens[0].includes(".")
      ? `${tokens[0]} ${tokens[1]}`
      : tokens[0];
  if (matchValue.length < 3 || GENERIC.has(matchValue)) {
    return null;
  }
  const normalizedSource = normalizeMerchant(raw);
  return {
    matchValue,
    operator: normalizedSource === matchValue ? "EQUALS" : "CONTAINS",
    display: titleCase(matchValue),
    field: "MERCHANT",
  };
}

export function splitMatchPhrases(raw: string): string[] {
  const seen = new Set<string>();
  const phrases: string[] = [];
  for (const part of raw.split(/[\n,;]+/)) {
    const value = part.trim().replace(/\s+/g, " ");
    if (value.length < 3 || value.length > 200) {
      continue;
    }
    const key = value.toLowerCase();
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    phrases.push(value);
  }
  return phrases;
}

export function merchantRuleSuggestion(input: {
  merchant: string | null | undefined;
  description: string | null | undefined;
  categoryCode: string | null | undefined;
  previousCategoryId: string | null | undefined;
  nextCategoryId: string;
  rules: MerchantRuleLike[];
}): MerchantRuleHint | null {
  if (input.categoryCode === "uncategorized" || input.previousCategoryId === input.nextCategoryId) {
    return null;
  }
  const merchant = input.merchant?.trim() || "";
  const description = input.description?.trim() || "";
  const merchantHint = stemMerchant(merchant);
  if (merchantHint && !fieldAlreadyCovered(merchant, "MERCHANT", input.rules)) {
    return { ...merchantHint, field: "MERCHANT" };
  }
  const descriptionHint = stemMerchant(description);
  if (descriptionHint && !fieldAlreadyCovered(description, "DESCRIPTION", input.rules)) {
    return { ...descriptionHint, field: "DESCRIPTION" };
  }
  return null;
}

export function fieldAlreadyCovered(source: string, field: string, rules: MerchantRuleLike[]): boolean {
  const haystack = normalizeMerchant(source);
  const expected = field.toUpperCase();
  return rules.some((rule) => {
    if (rule.enabled === false || rule.field.toUpperCase() !== expected) {
      return false;
    }
    const needle = normalizeMerchant(rule.matchValue);
    if (!needle) {
      return false;
    }
    switch (rule.operator.toUpperCase()) {
      case "EQUALS":
        return haystack === needle;
      case "STARTS_WITH":
        return haystack.startsWith(needle);
      default:
        return haystack.includes(needle);
    }
  });
}

export function merchantAlreadyCovered(source: string, rules: MerchantRuleLike[]): boolean {
  return fieldAlreadyCovered(source, "MERCHANT", rules);
}

export function formatMerchantRule(rule: MerchantRuleLike, categoryLabel: string): string {
  const field = rule.field.toUpperCase() === "DESCRIPTION" ? "Description" : "Merchant";
  const operator =
    rule.operator.toUpperCase() === "EQUALS"
      ? "equals"
      : rule.operator.toUpperCase() === "STARTS_WITH"
        ? "starts with"
        : "contains";
  return `${field} ${operator} “${rule.matchValue}” → ${categoryLabel}`;
}

function normalizeMerchant(raw: string): string {
  return raw
    .normalize("NFKC")
    .toLowerCase()
    .replaceAll("www.", " ")
    .replace(/[^\p{L}\p{N}.]+/gu, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function titleCase(value: string): string {
  return value
    .split(" ")
    .map((part) => (part.includes(".") ? part : part.charAt(0).toUpperCase() + part.slice(1)))
    .join(" ");
}

File: worthly-mobile/lib/theme/merchant_rule.dart
import 'package:worthly_mobile/api/models.dart';

class MerchantRuleHint {
  const MerchantRuleHint({
    required this.matchValue,
    required this.operator,
    required this.display,
    this.field = 'MERCHANT',
  });

  final String matchValue;
  final String operator;
  final String display;
  final String field;
}

const _prefixes = [
  'compra estrang',
  'compra mbw',
  'compra',
  'debito direto',
  'levantamento de numerario',
  'top-up by',
  'top up by',
];

const _drop = {
  'lda',
  'sa',
  'unipessoal',
  'comercial',
  'servicos',
  'serviços',
  'portugal',
  'lisboa',
  'porto',
  'braga',
  'coimbra',
  'faro',
  'setubal',
  'aveiro',
  'pt',
  'the',
  'de',
  'do',
  'da',
  'dos',
  'das',
  'e',
  'com',
  'www',
  'bill',
  'inst',
  'mb',
  'atm',
  'sucursal',
  'loja',
  'compra',
  'estrang',
  'mbw',
};

const _generic = {
  'paypal',
  'revolut',
  'klarna',
  'eupago',
  'mbway',
  'mb way',
  'trading 212',
  'trading212',
  'apple pay',
  'google pay',
  'compra',
};

MerchantRuleHint? stemMerchant(String? raw) {
  if (raw == null || raw.trim().isEmpty) {
    return null;
  }
  var text = _normalizeMerchant(raw);
  if (text.isEmpty) {
    return null;
  }
  var changed = true;
  while (changed) {
    changed = false;
    for (final prefix in _prefixes) {
      if (text == prefix || text.startsWith('$prefix ')) {
        text = text.substring(prefix.length).trim();
        changed = true;
      }
    }
  }
  final tokens = text
      .split(' ')
      .map((token) => token.replaceAll(RegExp(r'^\.+|\.+$'), ''))
      .where((token) => token.length >= 2 && !RegExp(r'^\d+$').hasMatch(token) && !_drop.contains(token))
      .toList();
  if (tokens.isEmpty) {
    return null;
  }
  final matchValue = tokens.length >= 2 && tokens.first.length <= 6 && !tokens.first.contains('.')
      ? '${tokens.first} ${tokens[1]}'
      : tokens.first;
  if (matchValue.length < 3 || _generic.contains(matchValue)) {
    return null;
  }
  final source = _normalizeMerchant(raw);
  return MerchantRuleHint(
    matchValue: matchValue,
    operator: source == matchValue ? 'EQUALS' : 'CONTAINS',
    display: _titleCase(matchValue),
  );
}

List<String> splitMatchPhrases(String raw) {
  final seen = <String>{};
  final phrases = <String>[];
  for (final part in raw.split(RegExp(r'[\n,;]+'))) {
    final value = part.trim().replaceAll(RegExp(r'\s+'), ' ');
    if (value.length < 3 || value.length > 200) {
      continue;
    }
    final key = value.toLowerCase();
    if (seen.contains(key)) {
      continue;
    }
    seen.add(key);
    phrases.add(value);
  }
  return phrases;
}

MerchantRuleHint? merchantRuleSuggestion({
  String? merchant,
  String? description,
  String? categoryCode,
  String? previousCategoryId,
  required String nextCategoryId,
  required List<CategorizationRule> rules,
}) {
  if (categoryCode == 'uncategorized' || previousCategoryId == nextCategoryId) {
    return null;
  }
  final merchantText = merchant?.trim() ?? '';
  final descriptionText = description?.trim() ?? '';
  final merchantHint = stemMerchant(merchantText);
  if (merchantHint != null && !fieldAlreadyCovered(merchantText, 'MERCHANT', rules)) {
    return MerchantRuleHint(
      matchValue: merchantHint.matchValue,
      operator: merchantHint.operator,
      display: merchantHint.display,
      field: 'MERCHANT',
    );
  }
  final descriptionHint = stemMerchant(descriptionText);
  if (descriptionHint != null && !fieldAlreadyCovered(descriptionText, 'DESCRIPTION', rules)) {
    return MerchantRuleHint(
      matchValue: descriptionHint.matchValue,
      operator: descriptionHint.operator,
      display: descriptionHint.display,
      field: 'DESCRIPTION',
    );
  }
  return null;
}

bool fieldAlreadyCovered(String source, String field, List<CategorizationRule> rules) {
  final haystack = _normalizeMerchant(source);
  final expected = field.toUpperCase();
  return rules.any((rule) {
    if (!rule.enabled || rule.field.toUpperCase() != expected) {
      return false;
    }
    final needle = _normalizeMerchant(rule.matchValue);
    if (needle.isEmpty) {
      return false;
    }
    switch (rule.operator.toUpperCase()) {
      case 'EQUALS':
        return haystack == needle;
      case 'STARTS_WITH':
        return haystack.startsWith(needle);
      default:
        return haystack.contains(needle);
    }
  });
}

bool merchantAlreadyCovered(String source, List<CategorizationRule> rules) {
  return fieldAlreadyCovered(source, 'MERCHANT', rules);
}

String formatMerchantRule(CategorizationRule rule, String categoryLabel) {
  final field = rule.field.toUpperCase() == 'DESCRIPTION' ? 'Description' : 'Merchant';
  final operator = switch (rule.operator.toUpperCase()) {
    'EQUALS' => 'equals',
    'STARTS_WITH' => 'starts with',
    _ => 'contains',
  };
  return '$field $operator “${rule.matchValue}” → $categoryLabel';
}

String _normalizeMerchant(String raw) {
  return raw
      .toLowerCase()
      .replaceAll('www.', ' ')
      .replaceAll(RegExp(r'[^\p{L}\p{N}.]+', unicode: true), ' ')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
}

String _titleCase(String value) {
  return value.split(' ').map((part) {
    if (part.contains('.') || part.isEmpty) {
      return part;
    }
    return '${part[0].toUpperCase()}${part.substring(1)}';
  }).join(' ');
}
